snuba/migrations/runner.py

```diff
@@ -359,3 +367,48 @@ def _get_migration_status(self) -> Mapping[MigrationKey, Status]:
                 raise e
 
         return data
+
+    @classmethod
+    def add_node(
+        self,
+        node_type: ClickhouseNodeType,
+        storage_sets: Sequence[StorageSetKey],
+        host_name: str,
+        port: int,
+        user: str,
+        password: str,
+        database: str,
+    ) -> None:
+        client_settings = ClickhouseClientSettings.MIGRATE.value
+        clickhouse = ClickhousePool(
+            host_name,
+            port,
+            user,
+            password,
+            database,
+            client_settings=client_settings.settings,
+            send_receive_timeout=client_settings.timeout,
+        )
+
+        migrations: List[MultiStepMigration] = []
+
+        for group in ACTIVE_MIGRATION_GROUPS:
+            group_loader = get_group_loader(group)
+
+            for migration_id in group_loader.get_migrations():
+                migration = group_loader.load_migration(migration_id)
+                migrations.append(migration)
+
+        for migration in migrations:
+            operations = (
+                migration.forwards_local()
+                if node_type == ClickhouseNodeType.LOCAL
+                else migration.forwards_dist()
+            )
+
+            for op in operations:
+                if isinstance(op, SqlOperation):
+                    if op._storage_set in storage_sets:
+                        sql = op.format_sql()
+                        print(f"Executing {sql}")
+                        clickhouse.execute(sql)
```

Comment:
I am not sure we can just rely on this behavior and skip non sql migrations. While we normally use python migrations for data migrations, this is not enforced in any way. We could totally have a python migration that generates a SQL statement that would have to be applied when starting up a new node.

Reply:
Yes, I think we need to change Python migrations to work properly for Distributed tables. I think we may need to assign them to a storage set and also possibly mark which ones need to be run in distributed vs local mode. For now, I think the simplest way is to skip them since we don't have any currently where this would be an issue.

Reply:
I don't think there's any good way to run Python migrations on a new node. Our Python migrations generally discover the relevant nodes by looking up the ClickHouse cluster configuration, and the new node by definition won't be added there yet. I think we need to either completely avoid having migrations in Python that need to be run to bring up the new node. Or RunPython needs to provide specific instructions on how to handle this scenario. e.g. it provides a special `execute_new_node` method, that accepts a ClickHouse host/port as argument and runs some function to bring just that node to the current state.

Reply:
What about this: in theory there is no reason for python migration not to be runnable when we set up a new node. The only obstacle I see is that, python migrations build up their own arbitrary clickhouse connection and generally rely on the cluster for doing so.
What if we provided a fake cluster object to them as a parameter instead of letting them build the connection?  This way we could provide a cluster that always returns connections to the new node irrespective on the cluster.

This way we would not need to figure out a way to skip python migrations when setting a node up. The sequence of migrations would be applied everywhere and we would be able to eventually also keep the migration state stored for each node.

Reply:
I think this is a scenario where the risk associated with accidentally executing the additional functions could be far greater that not doing anything.

Running the SQL operations on the new node is guaranteed to only ever mutate the state on one ClickHouse machine. Since that machine is the new one without any data on it yet, running the SQL operations is very low risk since it can't change any other parts of the system.

Since Python migrations are just functions, they can do anything. This is mostly intentional as we want the system to be capable of handling all of our future potential use cases. But it means they can mutate any parts of the system - any of the other ClickHouse nodes with important data, Redis, Kafka, any 3rd party services we might use now or in the future. 

Even if the migration got a fake cluster object, there's pretty much no guarantee that a specific Python migration is safe to rerun multiple times (by definition it's already been run at least once before) and also potentially rerun out of order, which would never otherwise happen.

So if we were choosing out of the 2 basic (simply skip or simply run), I would much prefer to err on the side of caution and skip the Python migrations. We could provide some hints or special instructions later around what else needs to be done to get the Python migration done on the new node.

I don't think this problem is one that needs to be solved immediately. We don't have any migrations that fall under this category (non SQL operation that needs to run when bringing up a new node) right now, and I can't think of one that would arise in the foreseeable future.

Ultimately, this command is supposed to be a tool to help manage multi node ClickHouse clusters, but manual intervention is still required when resizing clusters.

Reply:
After giving this a bit more thought, I decided to go ahead and add the `execute_new_node` method on a Python migration, allowing people to provide a separate method in case something else needs to be executed when a new node is added. We shouldn't just rerun the regular function as that may be dangerous.